lib/temporal/worker/activity_worker.rb

```diff
@@ -6,44 +6,46 @@ module Temporal
   class Worker
     class ActivityWorker
       def initialize(task_queue, core_worker, activities, converter, executor)
-        @running = false
         @task_queue = task_queue
         @worker = SyncWorker.new(core_worker)
         @activities = prepare_activities(activities)
         @converter = converter
         @executor = executor
         @running_activities = {}
         @cancellations = []
+        @shutdown_queue = Queue.new
       end
 
       def run(reactor)
-        @running = true
+        # @type var async_tasks: Array[Async::Task]
+        async_tasks = []
 
-        while running?
+        loop do
           activity_task = worker.poll_activity_task
-          reactor.async do
+          async_tasks << reactor.async do |async_task|
             if activity_task.start
               handle_start_activity(activity_task.task_token, activity_task.start)
             elsif activity_task.cancel
               handle_cancel_activity(activity_task.task_token, activity_task.cancel)
             end
+
+            async_tasks.delete(async_task)
           end
         end
+      rescue Temporal::Bridge::Error::WorkerShutdown
+        async_tasks.each(&:wait) # wait for all outstanding tasks to finish
+        shutdown_queue << nil
```

Comment:
We need to make sure shutdown happens on worker fatal error too just as if the user manually called shutdown

Reply:
Any exception from poll that isn't the shutdown one

Reply:
Will add this along with the changes to `Worker#run`, there lots of related ones